registry/remote/auth/client.go

```diff
@@ -365,3 +371,19 @@ func (c *Client) fetchOAuth2Token(ctx context.Context, realm, service string, sc
 	}
 	return "", fmt.Errorf("%s %q: empty token returned", resp.Request.Method, resp.Request.URL)
 }
+
+// rewindRequestBody tries to rewind the request body if exists.
+func rewindRequestBody(req *http.Request) error {
+	if req.Body == nil || req.Body == http.NoBody {
+		return nil
+	}
+	if req.GetBody == nil {
+		return fmt.Errorf("%s %q: request body is not rewindable", req.Method, req.URL)
+	}
+	body, err := req.GetBody()
```

Comment:
I would have expected this to actually call a seek. Is that possible without invoking getbody again or we ensuring that this is used only in the authentication flow?

Reply:
This is the common method to "seek" to be begining. The [transport.go](https://github.com/golang/go/blob/go1.17.6/src/net/http/transport.go#L659-L675) in the built-in golang package `net/http` does the same way.